### PR TITLE
Feature: thin plate spline

### PR DESCRIPTION
**Pull request by catalys1**
### Description
Implements thin plate spline warping, as per [872](https://github.com/kornia/kornia/issues/872).

### Status
**Work in progress**

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Breaking change (fix or new feature that would cause existing functionality to change)
- [x] New tests added to cover the changes
- [ ] Docstrings/Documentation updated

### PR Implementer
This is a small checklist for the implementation details of this PR.

If there are any questions regarding code style or other conventions check out our 
[summary](https://github.com/kornia/kornia/blob/master/CONTRIBUTING.rst).

- [x] Did you discuss the functionality or any breaking changes before ?
- [ ] **Pass all tests**: did you test in local ? `make test`
- [x] Unittests: did you add tests for your new functionality ?
- [ ] Documentations: did you build documentation ? `make build-docs`
- [x] Implementation: is your code well commented and follow conventions ? `make lint`
- [x] Docstrings & Typing: has your code documentation and typing ? `make mypy`
- [ ] Update notebooks & documentation if necessary

### KorniaTeam
<details>
  <summary>KorniaTeam workflow</summary>
  
  - [ ] Assign correct label
  - [ ] Assign PR to a reviewer
  - [ ] Does this PR close an Issue? (add `closes #IssueNumber` at the bottom if 
        not already in description)

</details>

### Reviewer
<details>
  <summary>Reviewer workflow</summary>

  - [ ] Do all tests pass? (Unittests, Typing, Linting, Documentation, Environment)
  - [ ] Does the implementation follow `kornia` design conventions?
  - [ ] Is the documentation complete enough ?
  - [ ] Are the tests covering simple and corner cases ?
 
</details>

### CODE REVIEW

**Comment by catalys1** [comment]
This is my first pull request, I'd appreciate it if someone could look it over and give me some feedback, let me know if I'm on the right track or not!

File: kornia/geometry/transform/thin_plate_spline.py [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+def _pair_square_euclidean(tensor1: torch.Tensor, tensor2: torch.Tensor) -> torch.Tensor:
+    """Compute the pairwise squared euclidean distance matrices (B, N, M) between two tensors
+        with shapes (B, N, C) and (B, M, C)."""
+    # ||t1-t2||^2 = (t1-t2)^T(t1-t2) = t1^T*t1 + t2^T*t2 - 2*t1^T*t2
+    t1_sq: torch.Tensor = tensor1.mul(tensor1).sum(dim=-1, keepdim=True)
+    t2_sq: torch.Tensor = tensor2.mul(tensor2).sum(dim=-1, keepdim=True).transpose(1, 2)
+    t1_t2: torch.Tensor = tensor1.matmul(tensor2.transpose(1, 2))
+    square_dist: torch.Tensor = -2 * t1_t2 + t1_sq + t2_sq
+    square_dist: torch.Tensor = square_dist.clamp(min=0)  # handle possible numerical errors
+    return square_dist
```
Comment (edgarriba) [inline kornia/geometry/transform/thin_plate_spline.py:26]: what about using `torch.cdist` ?

https://pytorch.org/docs/stable/generated/torch.cdist.html?highlight=distance

File: kornia/geometry/transform/thin_plate_spline.py [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+
+def _kernel_distance(squared_distances: torch.Tensor, eps: float = 1e-8) -> torch.Tensor:
+    """Compute the TPS kernel distance function: r^2*log(r), where r is the euclidean distance.
+    Since log(r) = 1/2*log(r^2), this function takes the squared distance matrix and calculates
+    0.5 * r^2 * log(r^2)."""
+    # r^2 * log(r) = 1/2 * r^2 * log(r^2)
+    return 0.5 * squared_distances * squared_distances.add(eps).log()
+
+
+def get_tps_parameters(points_src: torch.Tensor, points_target: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]:
```
Comment (edgarriba) [inline kornia/geometry/transform/thin_plate_spline.py:37]: I suggest a differents name according other functions we have - `get_tps_transform`. See https://kornia.readthedocs.io/en/latest/geometry.transform.html?highlight=get_parameters#kornia.geometry.transform.get_perspective_transform

For variable names `points_target` -> `points_dst`. We try to use `src` <-> `dst` convention.

File: kornia/geometry/transform/thin_plate_spline.py [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+    (B, N, 2); the second tensor contains the affine weights (B, 3, 2). In both cases, the last dimension contains
+    the weights for the x-transform and y-transform respectively.
+
+    Args:
+        points_src (torch.Tensor): batch of source points (B, N, 2) as (x, y) coordinate vectors
+        points_target (torch.Tensor): batch of target points (B, N, 2) as (x, y) coordinate vectors
+
+    Returns:
+        Tuple[torch.Tensor, torch.Tensor]: the (B, N, 2) kernel weights and (B, 3, 2) affine weights
+    """
```
Comment (edgarriba) [inline kornia/geometry/transform/thin_plate_spline.py:53]: for the docstring optional to add :math:`(B,N,2)` so that is rendered in latex style.

File: kornia/geometry/transform/thin_plate_spline.py [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+
```
Comment (edgarriba) [inline kornia/geometry/transform/thin_plate_spline.py:54]: Add type and shape check with meaningful messages.

File: kornia/geometry/transform/thin_plate_spline.py [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+                    kernel_weights: torch.Tensor, affine_weights: torch.Tensor) -> torch.Tensor:
+    r"""Warp a tensor of coordinate points using the thin plate spline defined by kernel points, kernel weights, and
+    affine weights.
+
+    The source points should be a (B, N, 2) tensor of (x, y) coordinates. The kernel centers are a (B, K, 2) tensor
+    of (x, y) coordinates. The kernel weights are a (B, K, 2) tensor, and the affine weights are a (B, 3, 2) tensor.
+    For the weight tensors, tensor[..., 0] contains the weights for the x-transform and tensor[..., 1] the weights
+    for the y-transform.
+
+    Returns a (B, N, 2) tensor containing the warped source points.
```
Comment (edgarriba) [inline kornia/geometry/transform/thin_plate_spline.py:89]: merge this to return section

File: kornia/geometry/transform/thin_plate_spline.py [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+
+    Args:
+        points_src (torch.Tensor): tensor of source points (B, N, 2)
+        kernel_centers (torch.Tensor): tensor of kernel center points (B, K, 2)
+        kernel_weights (torch.Tensor): tensor of kernl weights (B, K, 2)
+        affine_weights (torch.Tensor): tensor of affine weights (B, 3, 2)
```
Comment (edgarriba) [inline kornia/geometry/transform/thin_plate_spline.py:95]: missing end point sentence

File: kornia/geometry/transform/thin_plate_spline.py [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+        k_matrix[..., None].mul(kernel_weights[:, None]).sum(-2) +
+        points_src[..., None].mul(affine_weights[:, None, 1:]).sum(-2) +
+        affine_weights[:, None, 0]
+    )
+
+    return warped
+
+
+def warp_img_tensor_tps(tensor: torch.Tensor, kernel_centers: torch.Tensor, kernel_weights: torch.Tensor,
+                        affine_weights: torch.Tensor, align_corners: bool = False) -> torch.Tensor:
```
Comment (edgarriba) [inline kornia/geometry/transform/thin_plate_spline.py:116]: Function name too long I think. We could find something short and clean, what about `warp_image_tps` - it's close to opencv [`warpImage`](https://docs.opencv.org/3.4/df/dfe/classcv_1_1ShapeTransformer.html). We should check other libs how they name this.

File: kornia/geometry/transform/thin_plate_spline.py [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+    kernel weights, and affine weights.
+
+    The transform is applied to each pixel coordinate in the output image to obtain a point in the input
+    image for interpolation of the output pixel. So the TPS parameters should correspond to a warp from
+    output space to input space.
+
+    The input tensor is a (B, C, H, W) tensor. The kernel centers, kernel weight and affine weights are the
+    same as in ``warp_points_tps``.
+
+    Returns a warped image tensor with the same shape as the input.
```
Comment (edgarriba) [inline kornia/geometry/transform/thin_plate_spline.py:127]: merge with return section

File: kornia/geometry/transform/thin_plate_spline.py [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+
+    Args:
+        tensor (torch.Tensor): input image tensor (B, C, H, W)
```
Comment (edgarriba) [inline kornia/geometry/transform/thin_plate_spline.py:130]: end points

File: kornia/geometry/transform/thin_plate_spline.py [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+
+    Returns:
+        torch.Tensor: warped image tensor (B, C, H, W)
+    """
+    device, dtype = _extract_device_dtype([tensor, kernel_centers, kernel_weights, affine_weights])
+    batch_size, _, h, w = tensor.shape
+    ys, xs = torch.meshgrid(torch.linspace(-1, 1, h, device=device, dtype=dtype),
+                            torch.linspace(-1, 1, w, device=device, dtype=dtype))
+    coords: torch.Tensor = torch.stack([xs, ys], -1).view(-1, 2)
+    coords = torch.stack([coords] * batch_size, 0)  # expand to batch dimension
```
Comment (edgarriba) [inline kornia/geometry/transform/thin_plate_spline.py:143]: we have [kornia.utils.create_meshrid](https://kornia.readthedocs.io/en/latest/utils.html#kornia.utils.create_meshgrid)

File: test/geometry/transform/test_thin_plate_spline.py [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+
+import torch
+from torch.autograd import gradcheck
+from torch.testing import assert_allclose
+
+
+class TestTpsParameters:
+    @pytest.mark.parametrize("batch_size", [1, 2, 3])
+    @pytest.mark.parametrize("device", ('cpu', 'cuda'))
+    def test_smoke(self, batch_size, device):
```
Comment (edgarriba) [inline test/geometry/transform/test_thin_plate_spline.py:13]: you need to add `dtype` to the signature so that can be propagated throw the functional

File: test/geometry/transform/test_thin_plate_spline.py [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+            [1., -1.],
+            [0., 0.],
+        ]], device=device).repeat(batch_size, 1, 1)
+        dst = src.clone() * 2.
+        tensor = torch.zeros(batch_size, 3, 8, 8, device=device)
+        tensor[:, :, 2:6, 2:6] = 1.
+        expected = torch.ones_like(tensor)
+        kernel, affine = kornia.get_tps_parameters(dst, src)
+        warp = kornia.warp_img_tensor_tps(tensor, src, kernel, affine)
+        assert_allclose(warp, expected, atol=1e-4, rtol=1e-4)
```
Comment (edgarriba) [inline test/geometry/transform/test_thin_plate_spline.py:80]: tests are fine, you need to add more
- test_gradcheck
- test_jit
- test_exception (optional)

Check this to get a reference and namigs: https://github.com/kornia/kornia/blob/master/test/color/test_hsv.py#L12

File: kornia/geometry/transform/thin_plate_spline.py (hunk 7 above)
Comment (edgarriba) [reply kornia/geometry/transform/thin_plate_spline.py:116]: /cc @ducha-aiki @shijianjian @lferraz @anguelos @pmeier

File: kornia/geometry/transform/thin_plate_spline.py (hunk 1 above)
Comment (catalys1) [reply kornia/geometry/transform/thin_plate_spline.py:26]: I didn't use `torch.cdist` for two reasons: 1) I found errors in the gradient when I was using it a while back (it may be fixed now, not sure), and 2) there's no option to return the sqaured distance instead of the standard distance. But if you think it would be better, I could use `cdist` and square it instead.

File: kornia/geometry/transform/thin_plate_spline.py (hunk 7 above)
Comment (catalys1) [reply kornia/geometry/transform/thin_plate_spline.py:116]: I'm changing it to `warp_image_tps`. Any alternatives someone prefers?

File: kornia/geometry/transform/thin_plate_spline.py (hunk 1 above)
Comment (edgarriba) [reply kornia/geometry/transform/thin_plate_spline.py:26]: ok, that's fine I guess. @ducha-aiki any comment on that ?

File: test/geometry/transform/test_thin_plate_spline.py [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+import kornia.testing as utils
+
+import torch
+from torch.autograd import gradcheck
+from torch.testing import assert_allclose
+
+
+class TestTpsParameters:
+    @pytest.mark.parametrize("batch_size", [1, 2, 3])
+    @pytest.mark.parametrize("device", ('cpu', 'cuda'))
```
Comment (edgarriba) [inline test/geometry/transform/test_thin_plate_spline.py:12]: you don't need to add this by hand. Our test suite handles it automatically
